lms/djangoapps/certificates/signals.py

```diff
@@ -102,3 +104,29 @@ def _listen_for_passing_grade(sender, user, course_id, **kwargs):  # pylint: dis
             user=user.id,
             course=course_id
         ))
+
+
+@receiver(LEARNER_NOW_VERIFIED, dispatch_uid="learner_track_changed")
+def _listen_for_track_change(sender, user, **kwargs):  # pylint: disable=unused-argument
+    """
+    Catches a track change signal, determines user status,
+    fires downstream 'COURSE_GRADE_NOW_PASSED' task
+
+    """
+    if (
+        not waffle.waffle().is_enabled(waffle.SELF_PACED_ONLY) and
+        not waffle.waffle().is_enabled(waffle.INSTRUCTOR_PACED_ONLY)
+    ):
+        return
+    # This will read and update a user grade, if passed, will fire
+    # 'COURSE_GRADE_NOW_PASSED' signal downstream, including downstream
+    # waffle switch sorting for self paced/instructor paced.
+    user_enrollments = CourseEnrollment.enrollments_for_user(user=user)
+    grade_factory = CourseGradeFactory()
+    for enrollment in user_enrollments:
+        if grade_factory.read(user=user, course=enrollment.course).passed:
+            COURSE_GRADE_NOW_PASSED.send(
```

Comment:
This firing of a grade signal seems very odd to me.  It violates separation of concerns: the certificates layer firing a signal owned by the grades layer.
Also, by firing this signal, all other potential listeners of this grades signal would get triggered.
If I understand correctly, what you probably need to do is refactor the implementation of `_listen_for_passing_grade` and have this method call the same helper code.  Right?